### PR TITLE
feat: withdraw_fees gateway rpc command

### PR DESCRIPTION
**Pull request by Kodylow**
Adds a gateway rpc command to just get the withdraw fees to use when calculating a sweep transaction.

Partial fix to #2785

### CODE REVIEW

**Comment by elsirion** [comment]
Why does `gateway-cli` need to know that fee? Can't we tell `gatewayd` to send `all` and it then figures out internally what that means?

**Comment by elsirion** [comment]
The problem is (in rare occasions) the fee might change after substracting the fee from the withdraw amount, so it's not something you really want to do manually.

**Comment by Kodylow** [comment]
That makes sense, will update.

File: gateway/ln-gateway/src/rpc/mod.rs [hunk 1]
```diff
@@ -50,11 +51,31 @@ pub struct DepositAddressPayload {
...
     pub federation_id: FederationId,
-    #[serde(with = "bitcoin::util::amount::serde::as_sat")]
-    pub amount: bitcoin::Amount,
+    pub amount: AmountOrAll,
     pub address: Address,
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+pub enum AmountOrAll {
```
Comment (justinmoon) [inline gateway/ln-gateway/src/rpc/mod.rs:60]: Maybe call it `BitcoinAmountOrAll`? Just looking at the struct name I would assume this was a fedimint amount. Also, could you add a doc string to this struct https://x.com/dpc_pw/status/1722457120350589305?s=20

File: gateway/ln-gateway/src/lib.rs [hunk 2]
```diff
@@ -670,8 +672,13 @@ impl Gateway {
             address,
             federation_id,
         } = payload;
-
         let client = self.select_client(federation_id).await?;
+
+        let amount: bitcoin::Amount = match amount {
+            AmountOrAll::All => bitcoin::Amount::from_sat(client.get_balance().await.msats * 1000),
```
Comment (justinmoon) [inline gateway/ln-gateway/src/lib.rs:678]: I think this should probably use like `client.get_balance() - client.get_peg_out_fees()` and make sure this is valid amount

**Comment by justinmoon** [comment]
Might be good to also modify `fedimint-cli` to accept `all` as an amount?

**Comment by Kodylow** [comment]
@justinmoon see #3574 for add to fedimint-cli

File: gateway/ln-gateway/src/lib.rs [hunk 3]
```diff
@@ -670,12 +672,24 @@ impl Gateway {
...
 
-        let operation_id = client.withdraw(address, amount, fees).await?;
+        // TODO: Fees should probably be passed in as a parameter
+        let (amount, fee) = match amount {
+            // If the amount is "all", then we need to subtract the fees from
+            // the amount we are withdrawing
+            BitcoinAmountOrAll::All => {
+                let balance = bitcoin::Amount::from_sat(client.get_balance().await.msats * 1000);
+                let fees = client.get_withdraw_fee(address.clone(), balance).await?;
+                (balance - fees.amount(), fees)
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:684]: You should do this in a loop imo to check that the fee you have to pay stayed the same. There is no guarantee that the fee cannot rise from withdrawing a different amount (maybe there is currently due to our UTXO selection approach, but I wouldn't count on it).

File: gateway/ln-gateway/src/rpc/mod.rs [hunk 4]
```diff
@@ -50,11 +51,32 @@ pub struct DepositAddressPayload {
...
     pub address: Address,
 }
 
+/// Amount of bitcoin to send, or "all" to send all available funds
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(untagged)]
+pub enum BitcoinAmountOrAll {
+    Amount(#[serde(with = "bitcoin::util::amount::serde::as_sat")] bitcoin::Amount),
+    All,
+}
```
Comment (elsirion) [inline gateway/ln-gateway/src/rpc/mod.rs:64]: Do we want the serialization of `All` to be `null`? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=50dfc340d46f6d2652902b19a56ff6c1

File: gateway/ln-gateway/src/lib.rs (hunk 3 above)
Comment (dpc) [reply gateway/ln-gateway/src/lib.rs:684]: Oh, if the fee is supposed to be both of Fedimint tx and on chain fees, then we should use msats for `balance` (and the `fee`) BTW, just as I thought though.

Also - the loop termination condition should probably be "not increasing" or something. I suspect "stayed the same" can lead to hangs.

**Comment by Kodylow** [comment]
closing in favor of #3574

**Pull request closed without merging**